trino/trino.go

```diff
@@ -1253,13 +1263,403 @@ type queryResponse struct {
+	defer cancelFetch()
+
+	spoolingSegmentProcessors := 1
+	if numberOfSeg >= 2 {
+		spoolingSegmentProcessors = 2
+	}
+
+	var workersCloseSignals []chan struct{}
+
+	spooledSegmentStream := sp.processSegments(processedDataSegments, rowOffsetToSegmentIndex, totalRowsCount)
+
+	for range spoolingSegmentProcessors {
+		workersCloseSignals = append(workersCloseSignals, sp.proccessSpolledSegment(fetchContext, spooledSegmentStream, errorChannel, processedDataSegments, rowOffsetToSegmentIndex))
+	}
+
+	var fetchError error
+
+	finishedWorkersIndex := 0
+	for {
+		// to make sure that all works finish. Avoid leaking goroutines
+		if finishedWorkersIndex == len(workersCloseSignals) {
+			break
+		}
+
+		select {
+		case _, ok := <-workersCloseSignals[finishedWorkersIndex]:
+			if !ok {
+				workersCloseSignals[finishedWorkersIndex] = nil
+				finishedWorkersIndex++
+			}
+		case fetchError = <-errorChannel:
+			cancelFetch()
+		}
+	}
+
+	if fetchError != nil {
+		return nil, fetchError
+	}
+
+	combinedQueryData := make([]queryData, 0, int(totalRowsCount))
+
+	for _, segmentData := range processedDataSegments {
+		combinedQueryData = append(combinedQueryData, segmentData...)
+	}
+
+	return combinedQueryData, nil
+}
+
+// processSegments processes data segments and sends HTTP responses through a channel.
+//
+// It iterates over `sp.segments`, extracting metadata and handling each segment based on its type:
+// - "inline" segments are decoded and stored directly in `processedSegments`.
+// - "spooled" segments are fetched asynchronously, and their responses are sent to the output channel.
+//
+// The function also maps row offsets to segment indices for ordering.
+//
+// Errors encountered during processing are sent via the returned channel.
+//
+// Returns a receive-only channel (`<-chan httpResponseWithMeta`) that emits processed segment data or errors.
+func (sp *spoolingProtocol) processSegments(
+	processedSegments [][]queryData,
+	mappingSegmentsToArray map[int64]int,
+	dataSize int64,
+) <-chan spooledSegToProccess {
+	ch := make(chan spooledSegToProccess)
+	go func() {
+		defer close(ch)
+		for i, segment := range sp.segments {
+			segMap := segment.(map[string]interface{})
+			metaDataAssertion, _ := segMap["metadata"].(map[string]interface{})
+
+			metaData, err := parseSpoolingMetaData(metaDataAssertion)
+			if err != nil {
+				ch <- spooledSegToProccess{
+					err: err,
+				}
+			}
+
+			dataSize += metaData.RowsCount
+			mappingSegmentsToArray[metaData.RowOffset] = i
+
+			switch segMap["type"] {
+			case "inline":
+				data := segMap["data"].(string)
+				decodedBytes, err := base64.StdEncoding.DecodeString(data)
+				if err != nil {
+					ch <- spooledSegToProccess{
+						err: err,
+					}
+				}
+
+				decodedData, err := decode(decodedBytes, sp.encoder, metaData)
+				if err != nil {
+					ch <- spooledSegToProccess{
+						err: err,
+					}
+				}
+
+				processedSegments[i] = decodedData
```

Comment:
Why mutate the method parameter, instead of sending something to the channel? If this is correct, explain it in a code comment.